lib/wallet/src/Cardano/Wallet/Read/Primitive/Tx/Features/Integrity.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Copyright: © 2020-2022 IOHK
+-- License: Apache-2.0
+--
+
+module Cardano.Wallet.Read.Primitive.Tx.Features.Integrity
+    ( integrity
+    , txIntegrityF
+    , txIntegrity
+    , txIntegrityCardanoApi
+    )
+
+ where
+
+import Prelude hiding
+    ( (.) )
+
+import Cardano.Ledger.Alonzo.Tx
+    ( ScriptIntegrityHash )
+import Cardano.Ledger.Crypto
+    ( StandardCrypto )
+import Cardano.Ledger.SafeHash
+    ( SafeToHash (originalBytes) )
+import Cardano.Wallet.Read.Eras
+    ( EraFun (..), EraValue, K (..), applyEraFun, extractEraValue )
+import Cardano.Wallet.Read.Tx
+    ( Tx )
+import Cardano.Wallet.Read.Tx.Cardano
+    ( fromCardanoApiTx )
+import Cardano.Wallet.Read.Tx.Integrity
+    ( Integrity (..), getEraIntegrity )
+import Control.Category
+    ( (.) )
+import Data.Maybe.Strict
+    ( StrictMaybe, strictMaybeToMaybe )
+
+import qualified Cardano.Api as Cardano
+import qualified Cardano.Wallet.Primitive.Types.Hash as W
+
+integrity :: EraFun Integrity (K (Maybe (W.Hash "ScriptIntegrity")))
+integrity = EraFun
+    { byronFun = noIntegrity
+    , shelleyFun = noIntegrity
+    , allegraFun = noIntegrity
+    , maryFun = noIntegrity
+    , alonzoFun = yesIntegrity
+    , babbageFun = yesIntegrity
+    }
+    where
+        noIntegrity = const $ K Nothing
+        yesIntegrity (Integrity es) = K $ getIntegrity es
+
+getIntegrity
+    :: StrictMaybe (ScriptIntegrityHash StandardCrypto)
+    -> Maybe (W.Hash "ScriptIntegrity")
+getIntegrity = strictMaybeToMaybe . fmap (W.Hash . originalBytes)
+
+-- useful to cache this composition here
+txIntegrityF :: EraFun Tx (K (Maybe (W.Hash "ScriptIntegrity")))
+txIntegrityF = integrity . getEraIntegrity
+
+-- this is a helper that reuses the cache
+txIntegrity :: EraValue Tx -> Maybe (W.Hash "ScriptIntegrity")
+txIntegrity = extractEraValue . applyEraFun txIntegrityF
```

Comment:
```suggestion
-- this is a helper that reuses the cache
txIntegrity :: EraValue Tx -> Maybe (W.Hash "ScriptIntegrity")
txIntegrity = extractEraValue . applyEraFun txIntegrityF
  where
    txIntegrityF = integrity . getEraIntegrity
```

ℹ️ Just an informational remark: If you just want to reuse the cache without exporting it, you can use a local `where` binding.

It's also possible to inline — the operational semantics will be the preserved, and the cache only evaluated once. 🤓 In contrast, eta expansion will change the running time. In other words,

```hs
txIntegrity = extractEraValue . applyEraFun (integrity . getEraIntegrity)
```

and

```hs
txIntegrity = \tx -> extractEraValue $ applyEraFun (integrity . getEraIntegrity) tx
```

will treat the evaluation of `integrity . getEraIntegrity` differently.

Reply:
Yes, this was only an exercise to remind us how to cache and reuse the functions record. Nothing of this is used or should be exported for now. Removing